<center>
<h2>Capstone Project</h2>

# <span style='color: #81A5FF; '> 🥡 Extracting the Restaurants Menu's and Preprocessing them </span>

<h3> Group 4 </h3>

<span style='color: #081F4F; font-size: 18px;'>2023/2024</span>
</center>

This notebook concerns the preprocessement of the restaurant html code in order to extract their menu's, and further preprocessement. Essentially, the steps followed throughout this notebook are:

### Table of Contents <a class="anchor" id='toc'></a>

1. [🍽️ Data Import](#data-import)
2. [🍽️ Extracting the Relevant Part of the HTML code](#extracting-html)
3. [🍽️ Menu Translation](#menu-translation)



------

### 1. 🍽️ Data Import <a class="anchor" id="data-import"></a>
[Back to TOC](#toc)

In [1]:
import pandas as pd
import re
from langdetect import *
from deep_translator import  GoogleTranslator
from functions.menus import *
import pickle
import os

%load_ext autoreload
%autoreload 2

In [2]:
with open(f"menus_complete.jsonl", encoding="utf8") as f:
    lines = f.read().splitlines()

menus = pd.DataFrame(lines)
menus.columns = ['menu']

### 🍽️ 2. Extracting the Relevant Part of the Restaurant Page HTML code <a class="anchor" id="extracting-html"></a>
[Back to TOC](#toc)

> The data was scraped as HTML code of each restaurant's TheFork page. Now it is necessary to remove all unnecessary information such that there is only the URL of the restaurant page (to further merge with the rest of the data) and the respective menu. While the menu is being extracted, it is already being reformatted into a more readable format.

In [3]:
#Finding the URL
menus['input'] = menus['menu'].apply(extract_input)

In [5]:
#Extracting the menu section and reformatting it
menus['menu_pre_proc'] = menus['menu'].apply(lambda x: retrieve_menu(x))

In [6]:
menus.drop(['menu'], axis=1, inplace=True)

### 🍽️ 3. Menu Translation <a class="anchor" id="menu-translation"></a>
[Back to TOC](#toc)

> Now that we have the menus prepared, these will be translated both to portuguese and english. This is important because the product FlavourFlix is designed to cater portuguese cuisine to anyone, including tourists and non portuguese-speaking people. Nonetheless, its portuguese version is still relevant to maintain. The translations are done using the deep-translate package, resorting to the GoogleTranslator API.

In [ ]:
#Obtain the translations
menus['menus_en'] = menus['menus_pre_proc'].apply(lambda x: translate_menus(x, language='en'))
menus['menus_pt'] = menus['menus_pre_proc'].apply(lambda x: translate_menus(x, language='pt'))

In [33]:
menus.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1374 entries, 0 to 99
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   input          1373 non-null   object
 1   menu_pre_proc  1350 non-null   object
 2   menu_en        1342 non-null   object
 3   menu_pt        1342 non-null   object
dtypes: object(4)
memory usage: 53.7+ KB


> We find that only eight of the 1350 existing menus were not able to be translated, which is relatively good. Now we can export the data to incorporate it in the project.

In [34]:
menus.to_csv('menus_with_translations.csv', index=False)